# TensorFlow를 활용한 선형회귀 실습

이 노트북에서는 TensorFlow를 사용하여 간단한 선형회귀 문제를 풀어봅니다.

## 주요 단계
1. 데이터 생성 (y = 3x + 2 + 잡음)
2. 모델 정의 (Dense Layer)
3. 손실 함수 및 최적화 방법 설정
4. 학습 진행 및 시각화


In [1]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

print(tf.__version__)

2.21.0


## 1. 데이터 생성

In [2]:
# y = 3x + 2 + noise
n_samples = 5
X = np.linspace(-5, 5, n_samples)
y = 3 * X + 2 + np.random.normal(0, 1, n_samples)*20

plt.scatter(X, y)
plt.xlabel("X")
plt.ylabel("y")
plt.title("Generated Data")
plt.show()

/var/folders/76/lsnx8b5j45j73b66c6fjtg780000gn/T/ipykernel_24070/3034282863.py:10: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


## 2. 모델 정의

In [3]:
model = tf.keras.Sequential([
    tf.keras.layers.InputLayer(shape=(1,)),
    tf.keras.layers.Dense(1)
])

model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 1)              │             2 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2 (8.00 B)

 Trainable params: 2 (8.00 B)

 Non-trainable params: 0 (0.00 B)

## 3. 모델 컴파일
- 손실 함수: MSE (Mean Squared Error)
- 최적화 방법: SGD (Stochastic Gradient Descent)

In [4]:
model.compile(optimizer=tf.keras.optimizers.SGD(learning_rate=0.01),
              loss='mse')

## 4. 모델 학습

In [5]:
history = model.fit(X, y, epochs=200, verbose=2)

plt.plot(history.history['loss'])
plt.xlabel('Epoch')
plt.ylabel('Loss (MSE)')
plt.title('Training Loss')
plt.show()

Epoch 1/200


1/1 - 0s - 75ms/step - loss: 1106.4231


Epoch 2/200


1/1 - 0s - 13ms/step - loss: 761.0509


Epoch 3/200


1/1 - 0s - 12ms/step - loss: 566.5001


Epoch 4/200


1/1 - 0s - 12ms/step - loss: 456.7975


Epoch 5/200


1/1 - 0s - 12ms/step - loss: 394.8327


Epoch 6/200


1/1 - 0s - 13ms/step - loss: 359.7304


Epoch 7/200


1/1 - 0s - 13ms/step - loss: 339.7481


Epoch 8/200


1/1 - 0s - 14ms/step - loss: 328.2802


Epoch 9/200


1/1 - 0s - 12ms/step - loss: 321.6107


Epoch 10/200


1/1 - 0s - 13ms/step - loss: 317.6490


Epoch 11/200


1/1 - 0s - 13ms/step - loss: 315.2187


Epoch 12/200


1/1 - 0s - 13ms/step - loss: 313.6578


Epoch 13/200


1/1 - 0s - 12ms/step - loss: 312.5937


Epoch 14/200


1/1 - 0s - 13ms/step - loss: 311.8163


Epoch 15/200


1/1 - 0s - 13ms/step - loss: 311.2073


Epoch 16/200


1/1 - 0s - 12ms/step - loss: 310.6998


Epoch 17/200


1/1 - 0s - 12ms/step - loss: 310.2560


Epoch 18/200


1/1 - 0s - 13ms/step - loss: 309.8542


Epoch 19/200


1/1 - 0s - 12ms/step - loss: 309.4822


Epoch 20/200


1/1 - 0s - 13ms/step - loss: 309.1326


Epoch 21/200


1/1 - 0s - 13ms/step - loss: 308.8012


Epoch 22/200


1/1 - 0s - 12ms/step - loss: 308.4854


Epoch 23/200


1/1 - 0s - 12ms/step - loss: 308.1834


Epoch 24/200


1/1 - 0s - 13ms/step - loss: 307.8943


Epoch 25/200


1/1 - 0s - 12ms/step - loss: 307.6169


Epoch 26/200


1/1 - 0s - 13ms/step - loss: 307.3509


Epoch 27/200


1/1 - 0s - 13ms/step - loss: 307.0955


Epoch 28/200


1/1 - 0s - 12ms/step - loss: 306.8503


Epoch 29/200


1/1 - 0s - 13ms/step - loss: 306.6148


Epoch 30/200


1/1 - 0s - 12ms/step - loss: 306.3887


Epoch 31/200


1/1 - 0s - 13ms/step - loss: 306.1715


Epoch 32/200


1/1 - 0s - 12ms/step - loss: 305.9630


Epoch 33/200


1/1 - 0s - 13ms/step - loss: 305.7628


Epoch 34/200


1/1 - 0s - 12ms/step - loss: 305.5704


Epoch 35/200


1/1 - 0s - 12ms/step - loss: 305.3857


Epoch 36/200


1/1 - 0s - 12ms/step - loss: 305.2083


Epoch 37/200


1/1 - 0s - 13ms/step - loss: 305.0379


Epoch 38/200


1/1 - 0s - 13ms/step - loss: 304.8742


Epoch 39/200


1/1 - 0s - 13ms/step - loss: 304.7171


Epoch 40/200


1/1 - 0s - 13ms/step - loss: 304.5662


Epoch 41/200


1/1 - 0s - 12ms/step - loss: 304.4212


Epoch 42/200


1/1 - 0s - 12ms/step - loss: 304.2820


Epoch 43/200


1/1 - 0s - 12ms/step - loss: 304.1483


Epoch 44/200


1/1 - 0s - 12ms/step - loss: 304.0199


Epoch 45/200


1/1 - 0s - 12ms/step - loss: 303.8966


Epoch 46/200


1/1 - 0s - 12ms/step - loss: 303.7781


Epoch 47/200


1/1 - 0s - 12ms/step - loss: 303.6644


Epoch 48/200


1/1 - 0s - 12ms/step - loss: 303.5551


Epoch 49/200


1/1 - 0s - 13ms/step - loss: 303.4503


Epoch 50/200


1/1 - 0s - 12ms/step - loss: 303.3495


Epoch 51/200


1/1 - 0s - 13ms/step - loss: 303.2527


Epoch 52/200


1/1 - 0s - 12ms/step - loss: 303.1598


Epoch 53/200


1/1 - 0s - 13ms/step - loss: 303.0705


Epoch 54/200


1/1 - 0s - 12ms/step - loss: 302.9848


Epoch 55/200


1/1 - 0s - 12ms/step - loss: 302.9024


Epoch 56/200


1/1 - 0s - 12ms/step - loss: 302.8234


Epoch 57/200


1/1 - 0s - 13ms/step - loss: 302.7475


Epoch 58/200


1/1 - 0s - 13ms/step - loss: 302.6745


Epoch 59/200


1/1 - 0s - 12ms/step - loss: 302.6045


Epoch 60/200


1/1 - 0s - 11ms/step - loss: 302.5372


Epoch 61/200


1/1 - 0s - 13ms/step - loss: 302.4726


Epoch 62/200


1/1 - 0s - 13ms/step - loss: 302.4106


Epoch 63/200


1/1 - 0s - 12ms/step - loss: 302.3510


Epoch 64/200


1/1 - 0s - 13ms/step - loss: 302.2937


Epoch 65/200


1/1 - 0s - 13ms/step - loss: 302.2388


Epoch 66/200


1/1 - 0s - 12ms/step - loss: 302.1860


Epoch 67/200


1/1 - 0s - 12ms/step - loss: 302.1353


Epoch 68/200


1/1 - 0s - 13ms/step - loss: 302.0866


Epoch 69/200


1/1 - 0s - 12ms/step - loss: 302.0399


Epoch 70/200


1/1 - 0s - 12ms/step - loss: 301.9949


Epoch 71/200


1/1 - 0s - 13ms/step - loss: 301.9518


Epoch 72/200


1/1 - 0s - 12ms/step - loss: 301.9104


Epoch 73/200


1/1 - 0s - 12ms/step - loss: 301.8705


Epoch 74/200


1/1 - 0s - 13ms/step - loss: 301.8324


Epoch 75/200


1/1 - 0s - 13ms/step - loss: 301.7957


Epoch 76/200


1/1 - 0s - 12ms/step - loss: 301.7604


Epoch 77/200


1/1 - 0s - 12ms/step - loss: 301.7266


Epoch 78/200


1/1 - 0s - 12ms/step - loss: 301.6941


Epoch 79/200


1/1 - 0s - 12ms/step - loss: 301.6629


Epoch 80/200


1/1 - 0s - 13ms/step - loss: 301.6329


Epoch 81/200


1/1 - 0s - 12ms/step - loss: 301.6041


Epoch 82/200


1/1 - 0s - 13ms/step - loss: 301.5764


Epoch 83/200


1/1 - 0s - 13ms/step - loss: 301.5499


Epoch 84/200


1/1 - 0s - 12ms/step - loss: 301.5244


Epoch 85/200


1/1 - 0s - 13ms/step - loss: 301.4999


Epoch 86/200


1/1 - 0s - 12ms/step - loss: 301.4763


Epoch 87/200


1/1 - 0s - 12ms/step - loss: 301.4538


Epoch 88/200


1/1 - 0s - 12ms/step - loss: 301.4321


Epoch 89/200


1/1 - 0s - 14ms/step - loss: 301.4112


Epoch 90/200


1/1 - 0s - 13ms/step - loss: 301.3912


Epoch 91/200


1/1 - 0s - 12ms/step - loss: 301.3719


Epoch 92/200


1/1 - 0s - 12ms/step - loss: 301.3535


Epoch 93/200


1/1 - 0s - 12ms/step - loss: 301.3358


Epoch 94/200


1/1 - 0s - 12ms/step - loss: 301.3188


Epoch 95/200


1/1 - 0s - 13ms/step - loss: 301.3024


Epoch 96/200


1/1 - 0s - 13ms/step - loss: 301.2867


Epoch 97/200


1/1 - 0s - 12ms/step - loss: 301.2716


Epoch 98/200


1/1 - 0s - 12ms/step - loss: 301.2571


Epoch 99/200


1/1 - 0s - 12ms/step - loss: 301.2432


Epoch 100/200


1/1 - 0s - 13ms/step - loss: 301.2298


Epoch 101/200


1/1 - 0s - 12ms/step - loss: 301.2170


Epoch 102/200


1/1 - 0s - 12ms/step - loss: 301.2047


Epoch 103/200


1/1 - 0s - 12ms/step - loss: 301.1928


Epoch 104/200


1/1 - 0s - 13ms/step - loss: 301.1815


Epoch 105/200


1/1 - 0s - 12ms/step - loss: 301.1705


Epoch 106/200


1/1 - 0s - 13ms/step - loss: 301.1601


Epoch 107/200


1/1 - 0s - 12ms/step - loss: 301.1500


Epoch 108/200


1/1 - 0s - 12ms/step - loss: 301.1403


Epoch 109/200


1/1 - 0s - 12ms/step - loss: 301.1310


Epoch 110/200


1/1 - 0s - 12ms/step - loss: 301.1221


Epoch 111/200


1/1 - 0s - 13ms/step - loss: 301.1135


Epoch 112/200


1/1 - 0s - 13ms/step - loss: 301.1053


Epoch 113/200


1/1 - 0s - 12ms/step - loss: 301.0974


Epoch 114/200


1/1 - 0s - 12ms/step - loss: 301.0898


Epoch 115/200


1/1 - 0s - 12ms/step - loss: 301.0825


Epoch 116/200


1/1 - 0s - 13ms/step - loss: 301.0755


Epoch 117/200


1/1 - 0s - 14ms/step - loss: 301.0688


Epoch 118/200


1/1 - 0s - 15ms/step - loss: 301.0623


Epoch 119/200


1/1 - 0s - 12ms/step - loss: 301.0561


Epoch 120/200


1/1 - 0s - 13ms/step - loss: 301.0502


Epoch 121/200


1/1 - 0s - 13ms/step - loss: 301.0445


Epoch 122/200


1/1 - 0s - 12ms/step - loss: 301.0390


Epoch 123/200


1/1 - 0s - 13ms/step - loss: 301.0337


Epoch 124/200


1/1 - 0s - 12ms/step - loss: 301.0286


Epoch 125/200


1/1 - 0s - 12ms/step - loss: 301.0237


Epoch 126/200


1/1 - 0s - 12ms/step - loss: 301.0191


Epoch 127/200


1/1 - 0s - 12ms/step - loss: 301.0146


Epoch 128/200


1/1 - 0s - 13ms/step - loss: 301.0103


Epoch 129/200


1/1 - 0s - 12ms/step - loss: 301.0062


Epoch 130/200


1/1 - 0s - 12ms/step - loss: 301.0022


Epoch 131/200


1/1 - 0s - 12ms/step - loss: 300.9984


Epoch 132/200


1/1 - 0s - 13ms/step - loss: 300.9947


Epoch 133/200


1/1 - 0s - 13ms/step - loss: 300.9911


Epoch 134/200


1/1 - 0s - 12ms/step - loss: 300.9878


Epoch 135/200


1/1 - 0s - 13ms/step - loss: 300.9846


Epoch 136/200


1/1 - 0s - 13ms/step - loss: 300.9814


Epoch 137/200


1/1 - 0s - 13ms/step - loss: 300.9784


Epoch 138/200


1/1 - 0s - 12ms/step - loss: 300.9756


Epoch 139/200


1/1 - 0s - 12ms/step - loss: 300.9728


Epoch 140/200


1/1 - 0s - 12ms/step - loss: 300.9701


Epoch 141/200


1/1 - 0s - 12ms/step - loss: 300.9676


Epoch 142/200


1/1 - 0s - 13ms/step - loss: 300.9651


Epoch 143/200


1/1 - 0s - 12ms/step - loss: 300.9628


Epoch 144/200


1/1 - 0s - 13ms/step - loss: 300.9605


Epoch 145/200


1/1 - 0s - 12ms/step - loss: 300.9584


Epoch 146/200


1/1 - 0s - 13ms/step - loss: 300.9563


Epoch 147/200


1/1 - 0s - 13ms/step - loss: 300.9543


Epoch 148/200


1/1 - 0s - 12ms/step - loss: 300.9523


Epoch 149/200


1/1 - 0s - 13ms/step - loss: 300.9505


Epoch 150/200


1/1 - 0s - 13ms/step - loss: 300.9487


Epoch 151/200


1/1 - 0s - 12ms/step - loss: 300.9470


Epoch 152/200


1/1 - 0s - 13ms/step - loss: 300.9454


Epoch 153/200


1/1 - 0s - 12ms/step - loss: 300.9438


Epoch 154/200


1/1 - 0s - 12ms/step - loss: 300.9423


Epoch 155/200


1/1 - 0s - 12ms/step - loss: 300.9409


Epoch 156/200


1/1 - 0s - 12ms/step - loss: 300.9395


Epoch 157/200


1/1 - 0s - 13ms/step - loss: 300.9381


Epoch 158/200


1/1 - 0s - 12ms/step - loss: 300.9369


Epoch 159/200


1/1 - 0s - 12ms/step - loss: 300.9356


Epoch 160/200


1/1 - 0s - 12ms/step - loss: 300.9344


Epoch 161/200


1/1 - 0s - 13ms/step - loss: 300.9333


Epoch 162/200


1/1 - 0s - 12ms/step - loss: 300.9322


Epoch 163/200


1/1 - 0s - 12ms/step - loss: 300.9312


Epoch 164/200


1/1 - 0s - 12ms/step - loss: 300.9301


Epoch 165/200


1/1 - 0s - 13ms/step - loss: 300.9292


Epoch 166/200


1/1 - 0s - 13ms/step - loss: 300.9282


Epoch 167/200


1/1 - 0s - 12ms/step - loss: 300.9274


Epoch 168/200


1/1 - 0s - 13ms/step - loss: 300.9265


Epoch 169/200


1/1 - 0s - 12ms/step - loss: 300.9257


Epoch 170/200


1/1 - 0s - 12ms/step - loss: 300.9249


Epoch 171/200


1/1 - 0s - 13ms/step - loss: 300.9242


Epoch 172/200


1/1 - 0s - 12ms/step - loss: 300.9234


Epoch 173/200


1/1 - 0s - 12ms/step - loss: 300.9227


Epoch 174/200


1/1 - 0s - 12ms/step - loss: 300.9220


Epoch 175/200


1/1 - 0s - 13ms/step - loss: 300.9214


Epoch 176/200


1/1 - 0s - 12ms/step - loss: 300.9208


Epoch 177/200


1/1 - 0s - 12ms/step - loss: 300.9202


Epoch 178/200


1/1 - 0s - 12ms/step - loss: 300.9196


Epoch 179/200


1/1 - 0s - 12ms/step - loss: 300.9190


Epoch 180/200


1/1 - 0s - 13ms/step - loss: 300.9185


Epoch 181/200


1/1 - 0s - 12ms/step - loss: 300.9180


Epoch 182/200


1/1 - 0s - 11ms/step - loss: 300.9175


Epoch 183/200


1/1 - 0s - 12ms/step - loss: 300.9171


Epoch 184/200


1/1 - 0s - 12ms/step - loss: 300.9166


Epoch 185/200


1/1 - 0s - 12ms/step - loss: 300.9162


Epoch 186/200


1/1 - 0s - 13ms/step - loss: 300.9158


Epoch 187/200


1/1 - 0s - 13ms/step - loss: 300.9154


Epoch 188/200


1/1 - 0s - 12ms/step - loss: 300.9150


Epoch 189/200


1/1 - 0s - 13ms/step - loss: 300.9146


Epoch 190/200


1/1 - 0s - 13ms/step - loss: 300.9143


Epoch 191/200


1/1 - 0s - 12ms/step - loss: 300.9139


Epoch 192/200


1/1 - 0s - 12ms/step - loss: 300.9136


Epoch 193/200


1/1 - 0s - 12ms/step - loss: 300.9133


Epoch 194/200


1/1 - 0s - 12ms/step - loss: 300.9130


Epoch 195/200


1/1 - 0s - 11ms/step - loss: 300.9127


Epoch 196/200


1/1 - 0s - 13ms/step - loss: 300.9124


Epoch 197/200


1/1 - 0s - 12ms/step - loss: 300.9122


Epoch 198/200


1/1 - 0s - 12ms/step - loss: 300.9119


Epoch 199/200


1/1 - 0s - 13ms/step - loss: 300.9117


Epoch 200/200


1/1 - 0s - 12ms/step - loss: 300.9114


/var/folders/76/lsnx8b5j45j73b66c6fjtg780000gn/T/ipykernel_24070/4188173742.py:7: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


## 5. 결과 확인

In [6]:
W, b = model.layers[0].get_weights()
print(f"학습된 기울기 W: {W[0][0]:.3f}, 절편 b: {b[0]:.3f}")

y_pred = model.predict(X)

plt.scatter(X, y, label='Data')
plt.plot(X, y_pred, color='red', label='Fitted Line')
plt.legend()
plt.show()

학습된 기울기 W: 7.280, 절편 b: -4.133



1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 12ms/step


1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 17ms/step


/var/folders/76/lsnx8b5j45j73b66c6fjtg780000gn/T/ipykernel_24070/1249793878.py:9: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


## 생각해보기
- 학습 샘플 수가 작아지면 어떻게 되는가?

**나의 답변:**
샘플이 너무 적으면 몇 개 점에 직선이 맞춰져 참 관계(y = 3x + 2)를 못 찾고, 새 데이터에 대한 일반화가 나빠진다. 이번 실습도 샘플 5개만으로 참값 W=3, b=2 대신 W 약 7.28, b 약 -4.13이 나와 크게 어긋났다.


- 학습 샘플에 잡음이 많으면 어떻게 되는가? 이 경우 학습 데이터가 많이 필요한가?

**나의 답변:**
잡음이 크면 데이터가 참 직선 주변에 넓게 퍼져 MSE는 그 흩어진 점에 맞추려 하고, 이번처럼 손실이 약 301에서 멈추면 학습된 W, b가 참값과 멀어진다. 이 경우 학습 데이터를 많이 모아 잡음의 기댓값을 상쇄하는 것이 필요하다.
